### PR TITLE
Added Options to control "fuzzy" matching of userdata methods, etc

### PR DESCRIPTION
**Pull request by wledfor2**
Apologies for the resubmit. I rebased the wrong commit in my branch and destroyed the previous pull request (woops).

This pull request adds a new global option to control what (if any) fuzzy matching is done on fields/methods/properties/etc. For instance, now you can disable checking for userData:TestMethod() when Lua calls userData:testMethod().

This will slightly improve performance for users who don't use this feature. Additionally, this will allow us to easily add additional indexing functionality like #213 which I may also tackle in another pull request.

I re-implemented the previous pull request using an enum we use like a flag. The implementation would be a lot cleaner if we could use the Enum.HasFlag method, but unfortunately .net 3.5 did not have that.

In addition, I did not add new tests. When I wrote the original pull request I did not notice there were already tests for this (see Interop_TestNamesCamelized) -- which this passes.

### CODE REVIEW

File: src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace MoonSharp.Interpreter {
+
+	/// <summary>
+	/// A flag that controls if/how symbols (method, property, userdata) are fuzzily matched when they do not exist. Flags can be combined for multiple checks.
+	/// </summary>
+	[Flags]
+	public enum FuzzySymbolMatchingBehaviour
```
Comment (LimpingNinja) [inline src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: **Your updated pull fixes this** Slight** nitpick, C# uses American English spelling variations so does MoonSharp (see src/MoonSharp.Interpreter/Serialization/SerializationExtensions.cs). I would say keep it as "Behavior" for consistency sake.

File: src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+		/// <summary>The first letter of a symbol will be uppercased (to check for common C# naming conventions). For example, testMethod() becomes TestMethod()</summary>
+		UpperFirstLeter = 1,
+
+		/// <summary>Underscores in symbols are converted to camelcase. For example, test_method() becomes testMethod()</summary>
+		Camelify = 2,
+
+		/// <summary>
+		/// Combines both <see cref="UpperFirstLeter"/> and <see cref="Camelify"/>. For example, test_Method_two() becomes TestMethodTwo()
+		/// </summary>
+		All = UpperFirstLeter | Camelify
```
Comment (LimpingNinja) [inline src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: _**This is a discussion note, please feel free to argue/disagree**_ 

**I originally wrote this for your first change, but you fixed part of it to use it more like flags -- but the discussion point on ALL still sits here**

I understand what you were trying to do here, but I think this causes some logic problems.

Right now your code will do: 
1. Perform UpperFirstLetter
2. Perform Camelify
3. Perform CamelifyUpperFirstLetter (PascalCase)

In code though you are saying "All" which is only UpperFirstLetter and Camelify (according to the enum) but really means "PascalCase". In your version of the code it does PascalCase & UpperFirstLetter & Camelify inclusively, but won't allow an exclusive PascalCase run.

Also, if we later add other fuzzy methods too here, then all will become a bit more confusing in my opinion. For example, I just got a request for kebab-case and I want to add it so I change the enum:
```
KebabCase = 4
All = UpperFirstLetter | Camelify | Kebab-Case
```
You can see where this starts to get messy, because now if I just wanted to independently run:

1. UpperFirstLetter
2. Camelify
3. Kebab-Case

I automatically do a merge-run on all three as well, because the inclusive All flag.

In other words (and based on current code):
For example, PascalCase is NOT functionally equivalent to UpperFirstLetter | Camelify as those are independent; and there is no ability to run PascalCase with a fall-through on just Camelify while excluding just UpperFirstLetter.

```
None = 0
UpperFirstLetter = 1
Camelify = 2
PascalCase  = 4  // Or CamelifyUpperFirst
```

We don't really need an all as the _default behavior in code_ would be to run them all (exclusively) anyways if the flag is set and we use an &. If we want a mash-up we should define it like PascalCase. Or KebabCaseUpperFirst. That would allow us to easily use mixed-sets of evaluations in the case of fuzzy-matching gone wild.

Just playgrounding it, feel free to edit for discussion: https://dotnetfiddle.net/TP7OOX

File: src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+	}
+
+}
```
Comment (LimpingNinja) [inline src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: Ping.

File: src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs (hunk 1 above)
Comment (wledfor2) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: I assumed since there was a class named ColonMatchingBehaviour also used in the options, the project used British English. If will adjust this.

Comment (Benjamin-Dobell) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: Good catch, actually. We should actually fix that to be US spelling.

However, the rest of the project is US spelling (`Initialize`, `Prioritize` etc.) Generally speaking, I think it wise that all code just standardise on US spelling conventions; particularly in the case of C# as .NET itself uses US spelling. I'm not too fussed about comments, commit messages etc.

P.S. Yes, I know that above I spelt _standardise_ with British spelling; I'm an Aussie 😉

Comment (LimpingNinja) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: That's fair - looks like some inconsistent usage on our side; C# itself is based around american english, which could cause some weirdness in code; for example:

JsonSerializer serialiser = new JsonSerializer();
serialiser.Serialize(...)

With that said, since we are using Behaviour throughout the code right now, I'll create a separate issue to evaluate whether this should change - go ahead and follow the current pattern (Behaviour) for this.

Comment (LimpingNinja) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: @Benjamin-Dobell  commented while I was commenting - I was going to suggest allowing this through and spawning a new issue for checking other inconsistencies. What do you think?

Comment (Benjamin-Dobell) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: Sorry, @LimpingNinja.

As you were going to suggest; just leave it for now.

There's not a single occurrence of US spelling "Behavior" in the project; so we should just go with what's already in place i.e. leave it as `FuzzySymbolMatchingBehaviour`.

We can sort out the UK/US spelling issues in a future commit - bit of a pain for API exposed identifiers, however I do think it's worth it for consistency purposes. Would need to be a 2.x release etc.

Comment (wledfor2) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: Ackh. I committed my changes while you were speaking. Should I revert?

Comment (Benjamin-Dobell) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: I'll just be quiet and defer @LimpingNinja, rather than create any more confusion. Sorry!

Comment (LimpingNinja) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: Apologies not needed! Always good to get second responses on these. I'm agreed.

Note for future visitors, here's an example:

JsonSerializer serialiser = new JsonSerializer();
serialiser.Serialize(...)

Similary for things like Color, Initialze, etc. While Behaviour will likely have no direct correlate (I think only WPF has behavior-type classes) it is good to make sure that the mindset is already there for future consistency.

File: src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs (hunk 2 above)
Comment (wledfor2) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: I understand what you're saying. Originally it was labeled "Both" which didn't make sense, I guess "All" doesn't make any sense either (it defeats my purpose of adding more behavior in the future). So, remove All, add PascalCase, and set the default in ScriptGlobalOptions to UpperFirstLetter | Camelify | PascalCase. I believe this will address your concerns?

Comment (LimpingNinja) [reply src/MoonSharp.Interpreter/Options/FuzzySymbolMatchingBehaviour.cs]: That should do it!
